pkgs/jnigen/doc/lifecycle.md

```diff
@@ -0,0 +1,150 @@
+## Lifecycle of Java objects
+
+All Java classes generated by `package:jnigen` extend `JObject`. In JNI there
+are several kinds of references: local, global, and weak global. Local
+references are tied to a single thread. To enable seamless transfer of
+`JObject`s between isolates and safe usage in asynchronous code, `JObject`s
+always use **global** references. It's crucial to remember that there's a limit
+of approximately 50,000 global references.
+
+Once all references (in both Java and Dart) to an object are gone, Java's
+garbage collector (GC) can reclaim it.
+
+### Automatic reference management
+
+`JObject`s attach a native finalizer to their global references. Therefore, when
+the Dart GC collects them, the underlying Java reference is released.
+
+This approach works well for application code where a large number of references
+aren't created repeatedly. However, it's not ideal for packages because the
+usage patterns are unpredictable. This can lead to exceeding the global
```

Comment:
This could use some clarification that it's the package users that make it unpredictable:

"However, it's not ideal for packages because the usage patterns are unpredictable. You never know how your users will use your package."